File: testread.m
% Copyright 2020 by Lucata Corporation.
% Lucata grants a free license for noncommercial use and requires that this
% notice not be removed from the source files

% GraphBLAS C Test Suite
% DESCRIPTION: write matrices to input data directory

inputs

mkdir("data") ;
mkdir("data/testread") ;
mmwrite("data/testread/A.mm", A) ;
mmwrite("data/testread/B.mm", B) ;
mmwrite("data/testread/S.mm", S) ;
mmwrite("data/testread/M.mm", Mask) ;
mmwrite("data/testread/V1.mm", V1) ;
mmwrite("data/testread/V2.mm", V2) ;
mmwrite("data/testread/A_row.mm", A_row) ;
mmwrite("data/testread/A_col.mm", A_col) ;
mmwrite("data/testread/V1_ind.mm", V1_ind) ;
mmwrite("data/testread/V2_ind.mm", V2_ind) ;
mmwrite("data/testread/I_RANGE.mm", I_RANGE) ;
mmwrite("data/testread/I_STRIDE.mm", I_STRIDE) ;
mmwrite("data/testread/I_BACK.mm", I_BACK) ;
mmwrite("data/testread/ME.mm", ME) ;
mmwrite("data/testread/MEA.mm", MEA) ;
mmwrite("data/testread/MER.mm", MER) ;
mmwrite("data/testread/MES.mm", MES) ;
mmwrite("data/testread/MEB.mm", MEB) ;
mmwrite("data/testread/CE.mm", CE) ;
mmwrite("data/testread/CEA.mm", CEA) ;
mmwrite("data/testread/CER.mm", CER) ;
mmwrite("data/testread/CES.mm", CES) ;
mmwrite("data/testread/CEB.mm", CEB) ;
mmwrite("data/testread/VE.mm", VE) ;
mmwrite("data/testread/VEA.mm", VEA) ;
mmwrite("data/testread/VER.mm", VER) ;
mmwrite("data/testread/VES.mm", VES) ;
mmwrite("data/testread/VEB.mm", VEB) ;

File: testread.c
/*
   Copyright 2020 by Lucata Corporation.
   Lucata grants a free license for noncommercial use and requires that this
   notice not be removed from the source files.

   GraphBLAS C Test Suite
   DESCRIPTION: test reading and checking for all types
*/

#include "GraphBLAS.h"
#include "util/test_utils.h"

int main(int argc, char * argv[])
{
  GrB_Info info;
  OK(GrB_init(GrB_BLOCKING));
  testargs *myargs = get_test_args(argc, argv);
  bool testerror = false;

  printf("Running %s:\n", myargs->testbase); fflush(stdout);

  if (strlen(myargs->input0) == 0) strcpy(myargs->input0, "A");
  if (strlen(myargs->input1) == 0) strcpy(myargs->input1, "V1");
  if (strlen(myargs->input2) == 0) strcpy(myargs->input2, "V2");
  if (strlen(myargs->output) == 0) strcpy(myargs->output, "C");

  char ostr[256]; strcpy(ostr, myargs->output);
  for (int i = 0; i < num_Types(); i++) {
    info = GrB_SUCCESS;
    GrB_Type thetype; get_Type(i, &thetype);
    GrB_Matrix A = NULL;
    GrB_Vector V1 = NULL, V2 = NULL;

    read_matlab_matrix(myargs->testbase, myargs->input0, thetype, &A);
    read_matlab_vector(myargs->testbase, myargs->input1, thetype, &V1);
    read_matlab_vector(myargs->testbase, myargs->input2, thetype, &V2);

    sprintf(myargs->output, "%s_T%d", ostr, i);
    write_typed_matrix(myargs->testbase, myargs->output, A);
    testerror |= check_typed_matrix(myargs->testbase, myargs->output, A);

    sprintf(myargs->output, "%s%d_T%d", ostr, 1, i);
    write_typed_vector(myargs->testbase, myargs->output, V1);
    testerror |= check_typed_vector(myargs->testbase, myargs->output, V1);

    sprintf(myargs->output, "%s%d_T%d", ostr, 2, i);
    write_typed_vector(myargs->testbase, myargs->output, V2);
    testerror |= check_typed_vector(myargs->testbase, myargs->output, V2);

    OK (GrB_free (&A)); OK (GrB_free (&V1)); OK (GrB_free (&V2));
  }

  OK(GrB_finalize());
  return testerror;
}
